Add Type Hints, Fix Mypy Errors, and Refactor Code in AEPsych Directory

yalsaffar opened the pull request:
### Description:
This PR adds type hints, resolves Mypy errors, and refactors code across multiple modules in AEPsych. Key changes include type hint additions, tensor conversions for compatibility, and improved error handling. This is the first step towards resolving issue #366.

### Changes by Directory:

#### 1. **`acquisitions/`**:
- Added type hints across files.
- Refactored `construct_inputs_global_lookahead` to use tensors for `lb` and `ub` instead of lists.

#### 2. **`benchmark/`**:
- Added type hints.
- In `problem.py`, updated `LSEProblemWithEdgeLogging.evaluate` to use tensors for `metrics["prop_edge_sampling_err"]`.

#### 3. **`database/`, `generators/`, `kernels/`, `likelihoods/`**:
- Added type hints throughout.
- **`kernels/pairwisekernel.py`**: Updated `PairwiseKernel.forward` to avoid assigning tensor `d` as both an integer and tensor, ensuring Mypy compatibility.

#### 4. **`models/`**:
- Added type hints.
- **`gp_classification.py`**: Added assertions (`inducing_points is not None`) in the constructor and `_reset_variational_strategy`.
- **`semi_p.py`**: Added tensor conversion for `X` in `SemiParametricGPModel.posterior` to avoid Mypy errors.

#### 5. **Root-Level AEPsych Files**:
- Added type hints throughout.
- **`strategy.py`**: Added assertions to check for `None` in `self.model`, `self.x`, and `self.y` to prevent errors.
- **`utils.py`**: Wrapped all occurrences of `parname` with `str()` for compatibility with the `Config` class.
- **`plotting.py`**:
    - Refactored `plot_strat_3d` to improve handling of `contour_levels`:
        - Introduced detailed checks for `contour_levels_list` to ensure it is `Sized` (e.g., list or array).
        - Improved type handling for `slice_vals`, ensuring proper error messages and type checking.
        - Added error handling for cases where the model is `None`, raising a `RuntimeError` instead of silently failing.
    - Updated `plot_slice` to ensure better control over `contour_levels`:
        - Enforced type checking on `contour_levels` and ensured non-`None` values are provided, with clearer error messages.
        - Refactored the function to raise an explicit error (`ValueError`) if `contour_levels` is incorrectly set to `None`.

JasonKChow commented on aepsych/strategy.py:
```diff
@@ -350,10 +360,15 @@ def fit(self):
 
-    def update(self):
+    def update(self) -> None:
+        assert self.model is not None, "a model is needed here!"
```
Shouldn't be needed, see above.

JasonKChow commented on aepsych/strategy.py:
```diff
@@ -483,20 +500,20 @@ def _make_next_strat(self):
 
     @property
-    def finished(self):
+    def finished(self) -> Union[bool, Any]:
```
Should strictly be bool returned

JasonKChow commented on aepsych/strategy.py:
```diff
@@ -483,20 +500,20 @@ def _make_next_strat(self):
         return self._strat_idx == (len(self.strat_list) - 1) and self._strat.finished
 
-    def add_data(self, x, y):
+    def add_data(self, x, y) -> None:
```
Missing type hints

JasonKChow commented on aepsych/utils.py:
```diff
@@ -16,7 +16,7 @@
 
-def make_scaled_sobol(lb, ub, size, seed=None):
+def make_scaled_sobol(lb : Union[torch.Tensor, np.ndarray], ub : Union[torch.Tensor, np.ndarray], size: int, seed: Optional[int] = None) -> torch.Tensor:
     lb, ub, ndim = _process_bounds(lb, ub, None)
```
array/tensor question

yalsaffar replied:
It is true that `make_scaled_sobol` does not need this `Union[torch.Tensor, np.ndarray]`. However, `_process_bounds` and `promote_0d` methods still need to have this, as `_process_bounds` is used in almost the entire codebase. Fixing this will be the focus of #379 .

JasonKChow commented on aepsych/utils.py:
Missing typehints in a few functions here.

JasonKChow commented on aepsych/utils.py:
```diff
@@ -214,7 +214,7 @@ def get_jnd_multid(post_mean: torch.Tensor, mono_grid: torch.Tensor, df: int =1,
 
 
-def _get_ax_parameters(config):
+def _get_ax_parameters(config: Config):
```
Missing return types

JasonKChow commented on aepsych/utils.py:
```diff
@@ -238,15 +238,15 @@ def _get_ax_parameters(config):
     )
     choices = [
-        config.getlist(parname, "choices", element_type=str, fallback=["True", "False"])
+        config.getlist(str(parname), "choices", element_type=str, fallback=["True", "False"])
```
Are these string casts actually needed? It shouldn't.

JasonKChow replied:
That's fine. ax itself is also deprecated so in the future this function will be removed.

JasonKChow commented on aepsych/utils.py:
```diff
@@ -259,9 +259,10 @@ def _get_ax_parameters(config):
         try:
             try:
-                value = config.getfloat(parname, "value")
+                value: Union[float, str] = config.getfloat(str(parname), "value")
```
This should always be a float.

yalsaffar replied:
However, when the code reaches the following block:

```python
except ValueError:
    value = config.get(str(parname), "value")
```
The `value` could potentially be a string. Therefore, to handle this properly, we need to account for both possibilities. I'd be open to alternative approaches if there are better ways to resolve the mypy error related to this.

JasonKChow reviewed (changes requested):
Some typehints are missing:
from_config all over are missing return hints
lookahead_type
__init__ missing return None -- Semi_p.py file
problem.py Problem class properties missing return typehints.
problem.py has a TODO comment in sample_y method, maybe make a small PR to fix this if possible.
config.py Config missing type hints in methods (e.g., get_section())
db.py Database missing type hints
Likelihood semi_p.py file "expected_log_prob" missing type hints. 

Shouldn't everything be using Tensors now?

yalsaffar commented on aepsych/benchmark/problem.py:
```diff
@@ -121,7 +121,7 @@ def p_hat(self, model: aepsych.models.base.ModelProtocol) -> torch.Tensor:
     def evaluate(
         self,
-        strat: Union[Strategy, SequentialStrategy],
+        strat: SequentialStrategy,
```
Because some methods were only used for `SequentialStrategy`. But I fixed that with error handling now.

yalsaffar commented on aepsych/benchmark/problem.py:
```diff
@@ -212,7 +212,7 @@ class LSEProblem(Problem):
     """
 
-    def __init__(self, thresholds: Union[float, List]):
+    def __init__(self, thresholds: Union[float, List, torch.Tensor, None]) -> None:
```
I don't think so, but it is resolved now.

yalsaffar commented on aepsych/benchmark/problem.py:
```diff
@@ -316,10 +316,10 @@ def evaluate(self, strat: Union[Strategy, SequentialStrategy]) -> Dict[str, floa
         super().__init__(thresholds)
 
-    def evaluate(self, strat):
+    def evaluate(self, strat: SequentialStrategy) -> Dict[str, float]:
```
Just added it, but some conditions were needed to check the type and avoid calling `strat_list` for `Strategy`

yalsaffar commented on aepsych/models/ordinal_gp.py:
I noticed the missing `None` return type on the `__init__` method and tried to address it by adding the type hint. However, when I add any type hints to this constructor, I run into `mypy` errors like:
` error: "__init__" of "GPClassificationModel" gets multiple values for keyword argument
"covar_module"  [misc]
            super().__init__(`
and
`error: "__init__" of "GPClassificationModel" gets multiple values for keyword argument
"likelihood"  [misc]
            super().__init__(`

Even after using `pop` to remove the conflicting arguments from `kwargs`, the error persists. I'm not sure how to resolve this issue while keeping the type hints. Could you provide guidance on how to handle this?

JasonKChow replied:
Add type hints for arguments, then try again, if not ask again.

yalsaffar commented on aepsych/strategy.py:
```diff
@@ -310,7 +317,7 @@ def n_trials(self):
 
     def add_data(
```
This is the first function that would receive the data, as `np.ndarray` or `torch.Tensor`(as mentioned that in #403 )

yalsaffar commented on aepsych/strategy.py:
```diff
@@ -328,10 +335,12 @@ def add_data(
         if self.can_fit:
             if self.keep_most_recent is not None:
                 try:
+                    assert self.model is not None, "a model is needed here!"
```
Logically can_fit should be enough, but I still get the following error if I remove the assertion:
 `error: Item "None" of "AEPsychMixin | None" has no attribute "fit"  [union-attr] self.model.fit( `

JasonKChow replied:
Assert probably due to above Optional on can_fit. Wherever assert ends up being weirdly needed, it's likely because of an odd typing somewhere else that should be fixed instead.

yalsaffar commented:
I’ve gone ahead and added return type hints for the `from_config` method throughout the codebase. Along the way, I also fixed the type hints for `lookahead_type`, and made sure all the `__init__` methods that were missing `None` return types are now sorted out.

The TODO comment in `Problem.py` is still on the radar, but I’ll tackle that in a follow-up PR after this one. I also worked on `config.py` by fixing the missing type hints in methods like `get_section()`, and I cleaned up some issues in `db.py` related to the `Database`.

For `semi_p.py`, I added the missing type hints for `expected_log_prob`, even though it’s not currently being used. Most of the specific changes have already been resolved, but a few are still up for discussion. As soon as I get more feedback, I’ll add those as well.

Lastly, most parts of the code where tensors are needed for GPU support have been converted. The one exception is in the `add_data` method, which still accepts both tensors and `np.ndarrays`. This is because, as discussed in #403, it's the first point where data is accepted, and from there, everything is processed as tensors.

JasonKChow commented on aepsych/database/tables.py:
```diff
@@ -73,15 +75,15 @@ class DBMasterTable(Base):
 
     @classmethod
-    def from_sqlite(cls, row):
+    def from_sqlite(cls, row: Dict[str, Any]) -> 'DBMasterTable':
```
Please double check that search ahead types works without an extra import in all versions of Python we support (3.10-3.12).

yalsaffar replied:
Yes, I double-checked and it is supported in 3.7 and above.

JasonKChow commented on aepsych/generators/acqf_thompson_sampler_generator.py:
```diff
@@ -61,7 +61,7 @@ def __init__(
         self.stimuli_per_trial = stimuli_per_trial
 
-    def _instantiate_acquisition_fn(self, model: ModelProtocol):
+    def _instantiate_acquisition_fn(self, model: ModelProtocol) -> AnalyticExpectedUtilityOfBestOption:
```
Double check again. Doesn't appear changed.

JasonKChow commented on aepsych/generators/monotonic_rejection_generator.py:
```diff
@@ -63,7 +63,7 @@ def __init__(
         self.explore_features = explore_features
 
-    def _instantiate_acquisition_fn(self, model: MonotonicRejectionGP):
+    def _instantiate_acquisition_fn(self, model: MonotonicRejectionGP) -> MonotonicMCAcquisition:
```
Again, check if this is always correct.

JasonKChow commented on aepsych/models/gp_classification.py:
```diff
@@ -180,14 +181,15 @@ def _reset_hyperparameters(self):
             bounds=self.bounds,
             method=self.inducing_point_method,
         )
+        assert inducing_points is not None, "Inducing points cannot be None"
```
These are presumably for mypy errors, I don't like adding asserts to solve mypy errors, see if you can figure out a different way.

JasonKChow commented on aepsych/models/utils.py:
```diff
@@ -57,7 +58,7 @@ def select_inducing_points(
     bounds: Optional[Union[torch.Tensor, np.ndarray]] = None,
     method: str = "auto",
-):
+) -> Optional[torch.Tensor]:
```
Should always be tensors (also this Optional might be the cause of other problems above that require Asserts).

JasonKChow commented on aepsych/plotting.py:
```diff
@@ -356,7 +356,15 @@ def plot_strat_3d(
+    contour_levels_list: List[float] = []
+
+    if contour_levels is True:
+        contour_levels_list = [0.75]
```
It's a very complex argument it seems. It has to allow bool to default it to all ints. Double check if this is absolutely necessary.

JasonKChow commented on aepsych/strategy.py:
```diff
@@ -207,57 +207,63 @@ def normalize_inputs(
 
     @property
-    def finished(self):
+    def finished(self) -> Union[bool,torch.Tensor, None]:
```
This should be strictly bool, why not?

JasonKChow reviewed (changes requested):
Many comments. 

In general: avoid adding extra asserts to fix mypy problems, instead work out how to make the type hints align. Basically, no actual functional code should be written for most mypy problems, just modifying hints.

crasanders commented on aepsych/acquisition/lookahead.py:
```diff
@@ -406,12 +406,12 @@ def construct_inputs_global_lookahead(
     lookahead_type="levelset",
     target: Optional[float] = None,
     posterior_transform: Optional[PosteriorTransform] = None,
-    query_set_size: Optional[int] = 256,
```
This should still be optional. Line 415 needs to be changed.

crasanders commented on aepsych/acquisition/lookahead.py:
```diff
@@ -406,12 +406,12 @@ def construct_inputs_global_lookahead(
+) -> Dict[str, Any]:
+    lb = torch.tensor([bounds[0] for bounds in kwargs["bounds"]])
+    ub = torch.tensor([bounds[1] for bounds in kwargs["bounds"]])
     Xq = Xq if Xq is not None else make_scaled_sobol(lb, ub, query_set_size)
```
We need to pass in either Xq or query_set_size. We need to add a check that errors out if both are None.

yalsaffar replied:
Sure, I will add those checks after making `query_set_size` optional.

crasanders commented on aepsych/benchmark/problem.py:
```diff
@@ -316,22 +316,26 @@ def evaluate(self, strat: Union[Strategy, SequentialStrategy]) -> Dict[str, floa
-        n_opt_trials = strat.strat_list[-1].n_trials
+        if isinstance(strat, SequentialStrategy):
+            n_opt_trials = strat.strat_list[-1].n_trials
+        elif isinstance(strat, Strategy):
```
Is there ever a case where we pass in a Strategy instead of a SequentialStrategy?

yalsaffar replied:
You're right, but this is only achieved by removing the `Strategy` type hint from all `Problem` classes.

crasanders commented on aepsych/benchmark/problem.py:
```diff
@@ -316,22 +316,26 @@ def evaluate(self, strat: Union[Strategy, SequentialStrategy]) -> Dict[str, floa
         lb2 = lb + self.eps * r
         ub2 = ub - self.eps * r
 
+        assert isinstance(strat.x, torch.Tensor), "strat.x must be a tensor"
```
Shouldn't strat.x always be a tensor now that we don't accept numpy arrays anymore?

yalsaffar replied:
That's correct. However, `start.x` is initialized as `Optional[torch.Tensor]` since it starts as `None`. Therefore, this check is essential for mypy.

crasanders commented on aepsych/kernels/pairwisekernel.py:
```diff
@@ -40,11 +39,11 @@ def forward(self, x1, x2, diag=False, **params):
-            d = x1.shape[-1] - 1
-            assert d == x2.shape[-1] - 1, "tensors not the same dimension"
-            assert d % 2 == 0, "dimension must be even"
+            d_ = x1.shape[-1] - 1 #avoiding assigning d int then tensor(for mypy)
```
I'm confused. It looks like you just changed d to d_. Why does that matter to mypy?

yalsaffar replied:
`d` is assigned as both an `int` initially and a `tensor` at line 69. However, a better solution would be to define its type as `d: Union[torch.Tensor, int]`, which I’ll update in the code.

crasanders commented on aepsych/models/semi_p.py:
```diff
@@ -375,7 +376,9 @@ def predict(
 
-    def posterior(self, X, posterior_transform=None):
+    def posterior(self, X: Union[torch.Tensor, np.ndarray], posterior_transform: Optional[PosteriorTransform] = None) -> SemiPPosterior:
+        if isinstance(X, np.ndarray):
```
Is there a reason we still need to support numpy arrays in this method, or should we only allow tensors?

yalsaffar replied:
You're right; I found the root of the issue in `GPClassificationModel`, as it still supports both. I’ve fixed it where needed.

crasanders commented on aepsych/strategy.py:
```diff
@@ -207,57 +207,63 @@ def normalize_inputs(
-    def get_max(self, constraints=None, probability_space=False, max_time=None):
+    def get_max(self, constraints: Optional[Mapping[int, List[float]]]  = None, probability_space: bool = False, max_time: Optional[float] = None) -> Tuple[float, torch.Tensor]:
         constraints = constraints or {}
+        assert self.model is not None, "a model is needed here!"
```
I would make this message, and all the similar ones, more descriptive. Like "model is None! Cannot get the max without a model!"
